src/coreclr/jit/lowerarmarch.cpp

```diff
@@ -507,14 +506,55 @@ void Lowering::LowerCast(GenTree* tree)
 
     assert(!varTypeIsSmall(srcType));
 
-    if (tmpType != TYP_UNDEF)
+    // Transform:
+    //
+    //    *  CAST      long <- ulong <- uint                   *  AND      long
+    //    \--*  CAST      int <- small-uint <- int     =>      +--*  X        int
+    //       \--*  X         int                               \--*  CNS_INT  long  0xFF/0xFFFF
+    //
+    if (comp->opts.OptimizationEnabled() && op1->OperIs(GT_CAST) && !tree->isContained() && tree->TypeIs(TYP_LONG) &&
+        (dstType == TYP_ULONG))
```

Comment:
> `(dstType == TYP_ULONG)`

Is the intended check "is this cast zero-extending"? Then this should be `tree->IsUnsigned()`. `TYP_LONG/TYP_ULONG` distinction is meaningless for non-overflowing casts (which this outer cast will always be).